### PR TITLE
LoRA blog post

### PR DESCRIPTION
**Pull request by pcuenca**

### CODE REVIEW

File: _blog.yml [hunk 1]
```diff
@@ -1693,3 +1693,15 @@
     - cot
     - ift
     - sft
+
+- local: lora
+  title: "Using LoRA for Efficient Stable Diffusion Fine-Tuning"
+  author: pcuenq
+  thumbnail: /blog/assets/lora/thumbnail.png
+  date: January 25, 2023
```
Comment (sayakpaul) [inline _blog.yml:1701]: Change

File: lora.md [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+title: Using LoRA for Efficient Stable Diffusion Fine-Tuning
+thumbnail: /blog/assets/lora/thumbnail.png
+---
+
+<h1>
+	Using LoRA for Efficient Stable Diffusion Fine-Tuning
+</h1>
+
+<div class="blog-metadata">
+    <small>Published January 25, 2023.</small>
```
Comment (sayakpaul) [inline lora.md:11]: Change.

Comment (sayakpaul) [reply lora.md:11]: Also, we now host the images (except for the thumbnail) in `documentation-images`.

File: lora.md [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+    <a href="https://hf.co/sayakpaul">
+        <img class="avatar avatar-user" src="https://avatars.githubusercontent.com/u/22957388?v=4" width="100" title="Gravatar">
+        <div class="bfc">
+            <code>sayakpaul</code>
+            <span class="fullname">Sayak Paul</span>
+        </div>
+    </a>
+</div>
+
+[LoRA: Low-Rank Adaptation of Large Language Models](https://arxiv.org/abs/2106.09685) is a novel technique introduced by Microsoft researchers to deal with the problem of fine-tuning large-language models. Powerful models such as GPT-3, that contains 175B parameters, are prohibitevely expensive to fine-tune in order to adapt them to particular tasks or domains. LoRA proposes to freeze pre-trained model weights, but inject trainable layers in each transformer block. This greatly reduces the number of trainable parameters and GPU memory requirements, since gradients don't need to be computed for most of the model weights. The researchers found that by focusing on the Transformer attention blocks, fine-tuning quality with LoRA was on par with full model fine-tuning, while being much faster and requiring less compute.
```
Comment (sayakpaul) [inline lora.md:34]:
```suggestion
[LoRA: Low-Rank Adaptation of Large Language Models](https://arxiv.org/abs/2106.09685) is a novel technique introduced by Microsoft researchers to deal with the problem of fine-tuning large-language models. Powerful models such as GPT-3, that contain 175B parameters, are prohibitively expensive to fine-tune in order to adapt them to particular tasks or domains. LoRA proposes to freeze pre-trained model weights, but inject trainable layers in each transformer block. This greatly reduces the number of trainable parameters and GPU memory requirements, since gradients don't need to be computed for most of the model weights. The researchers found that by focusing on the Transformer attention blocks, fine-tuning quality with LoRA was on par with full model fine-tuning, while being much faster and requiring less compute.
```

Comment (sayakpaul) [reply lora.md:34]: Should we briefly mention?

* The layers contain low-rank matrices so that there's a good amount of savings in terms of memory. 
* These matrices are called update matrices and the blocks are often called adapter blocks.

File: lora.md [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+
+## LoRA for Diffusers
+
+Even though LoRA was initially proposed for large-language models, the technique can also be applied to other transformer-based models, or simply to models that contain transformer blocks. This is the case of Stable Diffusion, that uses transformer blocks to perform cross-attention between the image representation and the prompt that describes it. The details of the following figure (taken from the [stable diffusion paper](https://arxiv.org/abs/2112.10752)) are not important, just note that the yellow blocks are the ones in charge of building the relationship between image and text representations.
```
Comment (sayakpaul) [inline lora.md:38]:
```suggestion
Even though LoRA was initially proposed for large-language models, the technique can also be applied to other transformer-based models, or simply to models that contain transformer blocks. This is the case of Stable Diffusion, that uses transformer blocks to perform cross-attention between the image representation and the prompt that describes it. The details of the following figure (taken from the [Stable Diffusion paper](https://arxiv.org/abs/2112.10752)) are not important, just note that the yellow blocks are the ones in charge of building the relationship between image and text representations.
```

File: lora.md [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+        <div class="bfc">
+            <code>sayakpaul</code>
+            <span class="fullname">Sayak Paul</span>
+        </div>
+    </a>
+</div>
+
+[LoRA: Low-Rank Adaptation of Large Language Models](https://arxiv.org/abs/2106.09685) is a novel technique introduced by Microsoft researchers to deal with the problem of fine-tuning large-language models. Powerful models such as GPT-3, that contains 175B parameters, are prohibitevely expensive to fine-tune in order to adapt them to particular tasks or domains. LoRA proposes to freeze pre-trained model weights, but inject trainable layers in each transformer block. This greatly reduces the number of trainable parameters and GPU memory requirements, since gradients don't need to be computed for most of the model weights. The researchers found that by focusing on the Transformer attention blocks, fine-tuning quality with LoRA was on par with full model fine-tuning, while being much faster and requiring less compute.
+
+## LoRA for Diffusers
```
Comment (sayakpaul) [inline lora.md:36]: No strong opinions. 

```suggestion
## LoRA for Diffusers 🧨
```

File: lora.md [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+- **Trained weights are much, much smaller**. Because the original model is frozen and we inject new layers to be trained, we can save the weights for the new layers as a single file that weighs in at ~3 MB in size. This is about _one thousand times smaller_ than the original size of the UNet model!
+
+We are particularly excited about the last point. In order for users to share their awesome fine-tuned or _dreamboothed_ models, they had to share a full copy of the final model. Other users that want to try them out have to download the fine-tuned weights in their favorite UI, adding up to combined massive storage and download costs. As of today, there are about [1,000 Dreambooth models registered in the Dreambooth Concepts Library](https://huggingface.co/sd-dreambooth-library), and probably many more not registered in the library.
+
+With LoRA, it is now possible to publish [a single 3.29 MB file](https://huggingface.co/sayakpaul/sd-model-finetuned-lora-t4/blob/main/pytorch_lora_weights.bin) to allow others to use your fine-tuned model.
+
+_(h/t to [`@mishig25`](https://github.com/mishig25), the first person I heard use `dreamboothing` as a verb in a normal conversation)._
```
Comment (sayakpaul) [inline lora.md:56]: @mishig25 is the boss!

**Review by sayakpaul — APPROVED** [review]
So excited to ship it to prod 🚀 

Thanks a lot for working on it!

File: lora.md (hunk 4 above)
Comment (patrickvonplaten) [inline lora.md:38]: I'm not sure it's only limited to Transformer models no? One can essentially apply this method to all kinds of models I think no?

**Review by patrickvonplaten — APPROVED** [review]
Cool, easy-going language!

File: lora.md (hunk 3 above)
Comment (osanseviero) [reply lora.md:34]: Some additional nits

```suggestion
[LoRA: Low-Rank Adaptation of Large Language Models](https://arxiv.org/abs/2106.09685) is a novel technique introduced by Microsoft researchers to deal with the problem of fine-tuning large-language models. Powerful models such as GPT-3, which contains 175B parameters, are prohibitively expensive to fine-tune in order to adapt them to particular tasks or domains. LoRA proposes to freeze pre-trained model weights and inject trainable layers in each transformer block. This greatly reduces the number of trainable parameters and GPU memory requirements since gradients don't need to be computed for most model weights. The researchers found that by focusing on the Transformer attention blocks, fine-tuning quality with LoRA was on par with full model fine-tuning while being much faster and requiring less compute.
```

File: lora.md (hunk 4 above)
Comment (osanseviero) [reply lora.md:38]:
```suggestion
Even though LoRA was initially proposed for large-language models, the technique can also be applied to other transformer-based models or simply to models that contain transformer blocks. This is the case of Stable Diffusion, which uses transformer blocks to perform cross-attention between the image representation and the prompt that describes it. The details of the following figure (taken from the [Stable Diffusion paper](https://arxiv.org/abs/2112.10752)) are not important, just note that the yellow blocks are the ones in charge of building the relationship between image and text representations.
```

File: lora.md [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+
+## LoRA for Diffusers
+
+Even though LoRA was initially proposed for large-language models, the technique can also be applied to other transformer-based models, or simply to models that contain transformer blocks. This is the case of Stable Diffusion, that uses transformer blocks to perform cross-attention between the image representation and the prompt that describes it. The details of the following figure (taken from the [stable diffusion paper](https://arxiv.org/abs/2112.10752)) are not important, just note that the yellow blocks are the ones in charge of building the relationship between image and text representations.
+
+![Latent Diffusion Architecture](assets/lora/latent_diffusion.png)
+
+To the best of our knowledge, Simo Ryu ([`@cloneofsimo`](https://github.com/cloneofsimo)) was the first one to come up with a LoRA implementation adapted to Stable Diffusion. Please, do take a look at [their github project](https://github.com/cloneofsimo/lora) to see examples and lots of interesting discussions and insights.
```
Comment (osanseviero) [inline lora.md:42]:
```suggestion
To the best of our knowledge, Simo Ryu ([`@cloneofsimo`](https://github.com/cloneofsimo)) was the first one to come up with a LoRA implementation adapted to Stable Diffusion. Please, do take a look at [their GitHub project](https://github.com/cloneofsimo/lora) to see examples and lots of interesting discussions and insights.
```

File: lora.md [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+
+In order to inject LoRA trainable matrices as deep in the model as in the cross-attention layers, people used to need to hack the source code of [diffusers](https://github.com/huggingface/diffusers) in imaginative (but fragile) ways. If Stable Diffusion has showed us one thing, it is that the community always comes up with ways to bend and adapt the models for creative purposes, and we love that! Providing the flexibility to manipulate the cross-attention layers could be beneficial for many other reasons, such as making it easier to adopt optimization techniques such as [xFormers](https://github.com/facebookresearch/xformers). Other creative projects such as [Prompt-to-Prompt](https://arxiv.org/abs/2208.01626) could do with some easy way to access those layers, so we decided to [provide a general way for users to do it](https://github.com/huggingface/diffusers/pull/1639). We've been testing that _pull request_ since late December, and it's officially launching with our diffusers release today (v0.12).
```
Comment (osanseviero) [inline lora.md:44]:
```suggestion
In order to inject LoRA trainable matrices as deep in the model as in the cross-attention layers, people used to need to hack the source code of [diffusers](https://github.com/huggingface/diffusers) in imaginative (but fragile) ways. If Stable Diffusion has shown us one thing, it is that the community always comes up with ways to bend and adapt the models for creative purposes, and we love that! Providing the flexibility to manipulate the cross-attention layers could be beneficial for many other reasons, such as making it easier to adopt optimization techniques such as [xFormers](https://github.com/facebookresearch/xformers). Other creative projects such as [Prompt-to-Prompt](https://arxiv.org/abs/2208.01626) could do with some easy way to access those layers, so we decided to [provide a general way for users to do it](https://github.com/huggingface/diffusers/pull/1639). We've been testing that _pull request_ since late December, and it's officially launching with our diffusers release today (v0.12).
```

Comment (osanseviero) [inline lora.md:44]: Maybe link to 0.12 release notes or tweet

File: lora.md [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+
+We've been working with `@cloneofsimo` to provide LoRA training support in diffusers, for both Dreambooth and full fine-tuning methods! These techniques provide the following benefits:
```
Comment (osanseviero) [inline lora.md:46]: Maybe add link to cloneofsimo profile

File: lora.md [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+
+Full model fine-tuning of Stable Diffusion used to be slow and difficult, and that's part of the reason why lighter-weight methods such as Dreambooth or Textual Inversion have become so popular. With LoRA, it is much easier to fine-tune a model on a custom dataset.
+
+Diffusers now provides a [LoRA fine-tuning script](https://github.com/huggingface/diffusers/blob/main/examples/text_to_image/train_text_to_image_lora.py) that can run in as low as 11 GB of GPU RAM without resorting to tricks such as 8-bit optimizers. This is how you'd use it to fine-tune a model using [Lambda Labs Pokémon dataset](https://huggingface.co/datasets/lambdalabs/pokemon-blip-captions):
+
+```bash
+export MODEL_NAME="runwayml/stable-diffusion-v1-5"
+export OUTPUT_DIR="/sddata/finetune/lora/pokemon"
+export HUB_MODEL_ID="pokemon-lora"
+export dataset_name="lambdalabs/pokemon-blip-captions"
```
Comment (osanseviero) [inline lora.md:68]: nit for consistency
```suggestion
export DATASET_NAME="lambdalabs/pokemon-blip-captions"
```

File: lora.md [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+  --output_dir=${OUTPUT_DIR} \
+  --push_to_hub \
+  --hub_model_id=${HUB_MODEL_ID} \
+  --report_to=wandb \
+  --checkpointing_steps=500 \
+  --validation_prompt="Totoro" \
+  --seed=1337
+```
+
+One thing of notice is that the learning rate is `1e-4`, much larger than the usual learning rates for regular fine-tuning (in the order of `~1e-6`, typically). This is a [W&B dashboard](https://wandb.ai/pcuenq/text2image-fine-tune/runs/b4k1w0tn?workspace=user-pcuenq) of the previous run, which took about 5 hours in a 2080 Ti GPU (11 GB of RAM). I did not attempt to optimize the hyperparameters, so feel free to try it out yourself! [Sayak](https://huggingface.co/sayakpaul) did another run on a T4 (16 GB of RAM), here's [his final model](https://huggingface.co/sayakpaul/sd-model-finetuned-lora-t4), and here's [a demo Space that uses it](https://huggingface.co/spaces/pcuenq/lora-pokemon).
```
Comment (osanseviero) [inline lora.md:90]: To make space public https://huggingface.co/spaces/pcuenq/lora-pokemon

File: lora.md [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+
+```Python
+from huggingface_hub.repocard import RepoCard
+
+# LoRA weights ~3 MB
+model_path = "sayakpaul/sd-model-finetuned-lora-t4"
+
+card = RepoCard.load(model_path)
+model_base = card.data.to_dict()["base_model"]
+print(model_base)   # CompVis/stable-diffusion-v1-4
```
Comment (osanseviero) [inline lora.md:110]: Alternatively

```python
from huggingface_hub import model_info

info = model_info("sayakpaul/sd-model-finetuned-lora-t4")
print(info.cardData["base_model"]) #CompVis/stable-diffusion-v1-4
```

An advantage here is that no files are downloaded

File: lora.md [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+The above snippet will print the model he used for fine-tuning, which is `CompVis/stable-diffusion-v1-4`. In my case, I trained my model starting from version 1.5 of Stable Diffusion, so if you run the same code with [my LoRA model](https://huggingface.co/pcuenq/pokemon-lora) you'll see that the output is `runwayml/stable-diffusion-v1-5`.
+
+The information about the base model is automatically populated by the fine-tuning script we saw in the previous section, if you use the `--push_to_hub` option. This is recorded as a metadata tag in the `README` file of the model's repo, as you can see [here](https://huggingface.co/pcuenq/pokemon-lora/blob/main/README.md).
+
+After we determine the base model we used to fine-tune with LoRA, we load a normal Stable Diffusion pipeline. We'll customize it with the `DPMSolverMultistepScheduler` for very fast inference:
+
+```Python
+from diffusers import StableDiffusionPipeline, DPMSolverMultistepScheduler
+
+pipe = StableDiffusionPipeline.from_pretrained(model_base, torch_dtype=torch.float16)
```
Comment (osanseviero) [inline lora.md:122]: Also import torch above

**Review by osanseviero — APPROVED** [review]
Very cool blog post! :fire:

File: lora.md (hunk 9 above)
Comment (pcuenca) [reply lora.md:46]: It's linked the first time he's mentioned, will add it again here.

File: lora.md (hunk 12 above)
Comment (pcuenca) [reply lora.md:110]: Oh, that's simpler, thanks!

File: lora.md (hunk 4 above)
Comment (pcuenca) [reply lora.md]: You are right, I'm reading the paper and the sample code, and the technique can be applied elsewhere. I'll reword it a bit.
